### PR TITLE
Relay Spec

### PR DESCRIPTION
**Pull request by davimacedo**
This PR aims to make the Parse GraphQL API compliant with the [Relay Spec](https://relay.dev/docs/en/graphql-server-specification).

Close https://github.com/parse-community/parse-server/issues/5863

### CODE REVIEW

**Comment by omairvaiyani** [comment]
Good job man! Big PR - might be appropriate if you suggest a good way to split the reviewers for different files/sections so we can be more thorough

**Comment by davimacedo** [comment]
thanks @omairvaiyani it would be good if you guys could take a look and play around with the new schema. Then you could particularly review the changes related to the schema customization. But feel free to review any piece of code you want.

**Comment by Moumouls** [comment]
For my part, due to the massive changes, i'm going to start by a functionnal review with a live relay parse server.
I'll keep you informed.

**Comment by omairvaiyani** [comment]
I'm currently conduction a functional test only, as @douglasmuraoka has already covered the code review.

- [x] Loads up without any configuration or code changes from parse-server instance setup
- [x] Playground loads without errors
- [x] Playground displays schema and docs
- [ ] Schema contains types, fields, mutations based on ParseGraphQL configuration
- [ ] Can get single object with id
- [ ] Can update single object with id
- [ ] Can delete single object with id
- [ ] Can get multiple objects with constraints
- [ ] Can perform simple pagination
- [ ] Handles pointers for query constrains and mutations inputs
- [ ] Handles pointers for resolved fields
- [ ] Handles non-pointer array fields
- [ ] Returns default fields including ACLs

**Comment by davimacedo** [comment]
@omairvaiyani thanks for the review and feedback.

# clientMutationId

It is present in all mutations both in the input and in the output. The output value is always equal to the one passed in the input. Relay uses this field to uniquely identify the mutations.

You can find more information [here](https://relay.dev/graphql/mutations.htm)

# global id

It is actually according to the specification. You can find the regular Parse id in the objectId field, but the id field is used for global identification. It is basically a base64 string built by the combination of the class name and the object id. By this way you can globally identify an object and retrieve any object of any class by using this id through the `node` query. Relay uses this for refetching the objects.

You can find mode information [here](https://relay.dev/docs/en/graphql-server-specification#object-identification)

File: src/GraphQL/helpers/objectsQueries.js [hunk 1]
```diff
@@ -115,14 +163,126 @@ const findObjects = async (
...
+  }
+
+  return {
+    edges,
+    pageInfo,
+    count: preCount || count,
+  };
+};
+
+const calculateSkipAndLimit = (
```
[the remark this answers is not in the document]
Comment (davimacedo) [reply src/GraphQL/helpers/objectsQueries.js]: I've just added some comments and tests.

**Comment by omairvaiyani** [comment]
Got it, definitely not familiar enough with Relay spec to give more detailed feedback here. But in principle, the module is working on our codebase. Is there anything I can help to get this across the finish line?

**Comment by Moumouls** [comment]
The main tricky part of the relay implementation is the `cursor`, and it seems well implemented, i used the same logic in an old project.

I notice that in the SDL we have some `objectId` fields, i think we could remove now all reference to that field from the GraphQL API.

@davimacedo  so here we will only support the `Relay` version for the `GraphQL API` ?
I think it could be time consuming to maintain 2 versions of the `GQL API` in the future

On the other hand, we could write some examples in the `Parse GraphQL Doc` to help `GraphQL Beginners`

I'm going to use it (`Relay API`) intensely, in a project during the next 4 month.
I suggest to merge this PR, keep the `beta` notice on the `GraphQL` implementation.
If we do not notice to much bugs after the beta release we could remove the `beta` flag (may be beginning of 2020 #ANewYearForParse) ?

**Comment by Moumouls** [comment]
@davimacedo  And a huge big up for your tremendous work

**Comment by omairvaiyani** [comment]
@Moumouls 
> I notice that in the SDL we have some `objectId` fields, i think we could remove now all reference to that field from the GraphQL API.

Could you describe what, if any, limitations the Relay spec introduces vs the current implementation? For example, removing the `objectId` field, do you think that'll have any knock-on effects for users, or is using the `globalId` going to be relatively straightforward for existing Parse users?

The docs will essentially make or break our efforts here, if we're to release Relay asap, I suggest we start the docs asap.

**Comment by Moumouls** [comment]
@omairvaiyani  `Relay` do not introduce any limitations, it's a major upgrade in a `GraphQL API` life: 
- Better caching with `globalID` for front end clients (like `Apollo Client`)
- GlobalID: Allow further 
- Powerfull pagination system with `Cursors`
- Global normalization with `Connection` and `Node Interface`

**Currently `Relay Spec` is the most advanced/robust `GraphQL Extension Spec` that exists.**

To avoid a dual codebase on the `GraphQL` implementation i think we need to only keep one spec: 
- The `Custom` one inspired from the `Parse vocabulary`: for existing Parse Developers, I don't think it's going to attract new developers
- The `Relay` one, the best of `GraphQL Spec` on top of `Parse`: existing Parse Developers can learn new terms (with a short `tuto` on the docs), it can potentially attract new developers searching a powerfull and easy to use `GraphQL Server`

If we choose one and we choose to only keep the `Relay` implementation, we need to keep only one `ID System`: `Relay Global Id` 😄  (`objectId` will probably cause confusion)

@omairvaiyani  and yes you are right, we need to start a little tutorial on the `GraphQL Parse Docs` to help developers on boarding 👍 

Have I answered your questions? @omairvaiyani 

PS: `Global ID` is really easy to understand and to decode (juste a `Base64`) i think it's not a problem for developers. And if developers need assistance we can add a `Global ID Decoder` on the `parse-dashboard` to convert `GlobalID` to `objectID` ?

**Comment by omairvaiyani** [comment]
@Moumouls I'm completely with you in that using Relay will make Parse much more enticing to GraphQL developers. And I trust your opinion on robustness/stability. I still have some qualms about the omittance of `objectId` - mainly because current Parse developers will have extensively used it across their cloud codebase, as well as front-end codebase. If these developers were to make use of GraphQL on Parse, they will most like make an abstraction layer on their GraphQL client's to serialize/deserialize the `objectId`- so from that point of view, it shouldn't be a problem. It might be just an annoying extra-step when using the playground to not see the `objectId` anywhere, which is why I think we should maintain the field, even if it's not used to perform the queries.

**Comment by Moumouls** [comment]
@omairvaiyani  agree with let's keep the `objectId` field

**Comment by Moumouls** [comment]
Are we ready to merge this one @omairvaiyani  @davimacedo  ?

**Comment by omairvaiyani** [comment]
I'm happy with it, we could do with some directionality from the core-team around tackling the documentation

**Comment by Moumouls** [comment]
I started a documentation here for a complete release of the GraphQL implementation: https://github.com/parse-community/docs/pull/688

**Comment by davimacedo** [comment]
That's awesome guys! I think we are ready to merge. But it would be good to have the docs ready for the new Parse Server release (maybe 4.0?). I can help on this as well if you need @Moumouls. Sorry about the delay on answering this thread. I was pretty busy in the past few weeks but I should have more available time during December.

**Pull request merged**